File: Gold/15686.java
// 제목 : 치킨 배달
// 티어 : 골드 5
// 링크 : https://www.acmicpc.net/problem/15686
// 메모리(kb) : 14844
// 실행시간(ms) : 192

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

class Main
{
	static int answer;	//정답 == 도시의 치킨 거리
	static int n, m, M;	//도시의 크기, 살아남을 치킨집 수, 전체 치킨집 수
	static ArrayList<Integer[]> chickens;	//치킨집들 배열
	static ArrayList<Integer[]> houses;		//집들 배열
	static int[] chosenChickens;			//next_permutation에 m개의 치킨집 조합을 테스트해볼 배열
	
	static boolean next_permutation(int[] arr) {
		int len = arr.length;
		
		int i = len -1;
		while(i>0 && arr[i-1]>= arr[i]) --i;
		if(i==0) return false;
		
		int j = len-1;
		while(arr[i-1]>=arr[j]) --j;
		
		swap(arr, i-1,j);
		
		int k = len-1;
		while(i<k) {
			swap(arr,i,k);
			i++;
			k--;
		}
		
		return true;
	}
	
	static void swap(int[]arr, int i, int j) {
		int temp = arr[i];
		arr[i] = arr[j];
		arr[j] = temp;
	}
	
	//현재 치킨집 조합으로 도시의 치킨 거리를 구하고, 다른 조합에서 나온 최솟값과 비교하는 함수
	public static void cityDistance() {
		int cur_dis = 0;
		
		for(Integer[] house : houses) cur_dis += houseDistance(house);
		
		answer = Math.min(answer, cur_dis);
	}
	
	//한 집에서 남은 치킨집들과 치킨 거리를 구하는 함수
	public static int houseDistance(Integer[] house) {
		int distance = Integer.MAX_VALUE;
		
		for(int i=0;i<M;i++) {
			if(chosenChickens[i]==1) {
				Integer[] chicken = chickens.get(i);
				int cur_dis = Math.abs(house[0]-chicken[0])+Math.abs(house[1]-chicken[1]);
				distance = Math.min(cur_dis, distance);
			}
		}
		
		return distance;
	}
	
    public static void main(String args[]) throws Exception
    {
    	
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        
        //전역 초기화
        answer = Integer.MAX_VALUE;
        chickens = new ArrayList<>();
        houses = new ArrayList<>();

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        M = 0;
        
        //도시 지도를 읽으면서 집과 치킨집을 기록
        for(int i=0;i<n;i++) {
        	st = new StringTokenizer(br.readLine());
        	for(int j=0;j<n;j++) {
        		int a = Integer.parseInt(st.nextToken());
        		if(a==1) houses.add(new Integer[]{i,j});
        		if(a==2) {
        			chickens.add(new Integer[]{i,j});
        			++M;
        		}
        	}
        }

        //m개를 뽑는 조합을 위해 chosenChickens을 만들고, 1을 끝에서부터 m개 넣음
        chosenChickens = new int[M];
        for(int i=0; i<m;i++) chosenChickens[M-1-i]=1;
        
        //도시의 최소 치킨 거리 찾기
        do {
        	cityDistance();
        	
        }while(next_permutation(chosenChickens));
        
        //출력
        System.out.println(answer);
    }
}

File: Gold/15686.cpp
// 제목 : 치킨 배달
// 티어 : 골드 5
// 링크 : https://www.acmicpc.net/problem/15686
// 메모리(kb) : 2024
// 실행시간(ms) : 4

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

//선택받은 m개의 치킨집 목록과, 가정집 목록을 매개변수로 받아 치킨거리를 구하는 함수
int getSum(vector<pair<int, int>> mChicken, vector<pair<int, int>> house) {
	//도시의 치킨거리
	int sum = 0;
	//각각의 집마다
	for (int i = 0; i < house.size(); i++) {
		//비교하며 최솟값을 구할 치킨거리
		int tmp = 1e9;
		int hx = house[i].first;
		int hy = house[i].second;
		//치킨집들과의 치킨거리를 구하며 최솟값으로 갱신
		for (int j = 0; j < mChicken.size(); j++) {
			int mx = mChicken[j].first;
			int my = mChicken[j].second;
			int mLength = abs(hx - mx) + abs(hy - my);
			tmp = min(tmp, mLength);
		}
		//도시의 치킨거리에 더함
		sum += tmp;
	}
	return sum;
}

int main() {
	//최대 크기로 잡은 도시 지도와 도시의 크기
	int arr[51][51];
	int n, m;
	//집, 치킨집 목록
	vector<pair<int, int>> house;
	vector<pair<int, int>> chicken;
	cin >> n >> m;
	
	//지도에 저장하면서 가정집과 치킨집이 있다면 목록에 저장하기
	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= n; j++) {
			cin >> arr[i][j];
			if (arr[i][j] == 1) house.push_back({ i,j });
			if (arr[i][j] == 2) chicken.push_back({ i,j });
		}
	}


	//next_permutation에 넣고 돌려서 m개의 치킨집을 구해줄 벡터
	vector<bool> permuHelper;
	for (int i = 0; i < (chicken.size() - m); i++) {
		permuHelper.push_back(false);
	}
	for (int i = 0; i < m; i++) {
		permuHelper.push_back(true);
	}

	int result = 1e9;

	//치킨집을 m개 고르는 각각의 조합마다 다 시험
	do {
		//선택받은 m개의 치킨집 목록
		vector<pair<int, int>> mChicken;
		for (int i = 0; i < chicken.size(); i++) {
			if (permuHelper[i] == true) {
				int mx = chicken[i].first;
				int my = chicken[i].second;
				mChicken.push_back({ mx,my });
			}
		}
		//최솟값과 비교
		result = min(result, getSum(mChicken, house));
	} while (next_permutation(permuHelper.begin(), permuHelper.end()));

	//출력
	cout << result;
}